src/generator.rs

```diff
@@ -107,3 +124,82 @@ impl<'a> Generator<'a> {
         }
     }
 }
+
+#[cfg(test)]
+pub mod test {
+    use crate::MutationError;
+    use rand::SeedableRng;
+    use rand::prelude::SliceRandom;
+
+    /// Test the error when there are no possible choices.
+    #[test]
+    pub fn test_generator_no_choices() {
+        let mut gen = crate::generator::Generator::from(&[0, 1, 2, 3, 4, 5][..]);
+        let choices: [usize; 0] = [];
+        assert_eq!(gen.choose(&choices, "error"), Err(MutationError::Check("Empty choices".to_string())));
+    }
+
+    /// Test that a generator can be used to choose one value.
+    #[test]
+    pub fn test_generator_one_choices() {
+        let mut gen = crate::generator::Generator::from(&[0, 1, 2, 3, 4, 5][..]);
+        let choices = [0];
+        assert_eq!(gen.choose(&choices, "error").unwrap(), &0);
+    }
+
+    /// Test that a generator can be used to choose between two values.
+    #[test]
+    pub fn test_generator_choose_two_values() {
+        let mut gen = crate::generator::Generator::from(&[0, 1, 2, 3, 4, 5][..]);
+        let choices = [0, 1];
+        assert_eq!(gen.choose(&choices, "error").unwrap(), &0);
+        assert_eq!(gen.choose(&choices, "error").unwrap(), &1);
+        assert_eq!(gen.choose(&choices, "error").unwrap(), &0);
+        assert_eq!(gen.choose(&choices, "error").unwrap(), &1);
+        assert_eq!(gen.choose(&choices, "error").unwrap(), &0);
+        assert_eq!(gen.choose(&choices, "error").unwrap(), &1);
+
+        // This is the only case where we can't choose a value, because we have 2 choices and 6 bytes.
+        assert_eq!(gen.choose(&choices, "error"), Err(MutationError::Check("Ran out of entropy".to_string())));
+    }
+
+    /// Test that a generator can be used to choose between three values.
+    #[test]
+    pub fn test_generator_choose_three_values() {
+        let mut gen = crate::generator::Generator::from(&[0, 1, 2, 3, 4, 5][..]);
+        let choices = [0, 1, 2];
+        assert_eq!(gen.choose(&choices, "error").unwrap(), &0);
+        assert_eq!(gen.choose(&choices, "error").unwrap(), &1);
+        assert_eq!(gen.choose(&choices, "error").unwrap(), &2);
+        assert_eq!(gen.choose(&choices, "error").unwrap(), &0);
+        assert_eq!(gen.choose(&choices, "error").unwrap(), &1);
+        assert_eq!(gen.choose(&choices, "error").unwrap(), &2);
+
+        // This is the only case where we can't choose a value, because we have 3 choices and 6 bytes.
+        assert_eq!(gen.choose(&choices, "error"), Err(MutationError::Check("Ran out of entropy".to_string())));
+    }
```

Comment:
A lot of this is testing the implementation details of `arbitrary`, aside from the final check of the `MutationError`. You could keep future such tests simpler.

Reply:
@maackle well, that's only because in this case the method being tested is a thin wrapper around arbitrary, but in the case of a future refactor it might do something else

FYI these tests already found a bug in the implementation (it would repeatedly choose the same value when it runs out of entropy and silently fail), so i'm not sure they could have been any simpler to be safe

Reply:
also FYI the tests were mostly autogenerated by copilot, so it wasn't a big lift to flesh them out